### PR TITLE
refactor: call opcodes

### LINKED ISSUE #646: feat(gas): charge create gas costs in `call` opcodes
### Describe the Feature Request

When executing an opcode from the `CALL` family, the access cost gas must be charged according to if it's a warm or a cold access.

### LINKED ISSUE #647: refactor: delete `prepare_call` helper
### Describe the Feature Request

With the complexity added by gas calculations, having a `prepare_call` helper complexifies the codebase for little benefits. In this refactor, you will remove it and prefer to "prepare" the call directly inside the opcode

### PR DESCRIPTION
**Pull request by enitrat**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #645 Resolves #647 Resolves #646

## What is the new behavior?

- Refactored call opcodes by deleting prepare_call 
- Added message_call gas accounting
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/648)

### CODE REVIEW

File: crates/evm/src/call_helpers.cairo [hunk 1]
```diff
@@ -44,108 +45,57 @@ enum CallType {
...
-            }
-        )
-    }
-
     /// Initializes and enters into a new sub-context
     /// The Machine will change its `current_ctx` to point to the
     /// newly created sub-context.
     /// Then, the EVM execution loop will start on this new execution context.
-    fn generic_call(ref self: VM, call_args: CallArgs) -> Result<(), EVMError> {
+    fn generic_call(
```
Comment (bajpai244) [inline crates/evm/src/call_helpers.cairo:52]: We should use naming of variables as per ELS wherever possible.

image

File: crates/evm/src/call_helpers.cairo [hunk 2]
```diff
@@ -44,108 +45,57 @@ enum CallType {
...
         if self.message().depth >= constants::STACK_MAX_DEPTH {
-            self.gas_left += call_args.gas;
+            self.gas_left += gas;
             return self.stack.push(0);
         }
 
+        let mut calldata = Default::default();
+        self.memory.load_n(args_size, ref calldata, args_offset);
+
         // Case 2: `to` address is not a precompile
```
Comment (bajpai244) [inline crates/evm/src/call_helpers.cairo:75]: Is this comment correctly placed?

File: crates/evm/src/call_helpers.cairo [hunk 3]
```diff
@@ -44,108 +45,57 @@ enum CallType {
...
+        let caller = Address {
+            evm: caller, starknet: kakarot_core.compute_starknet_address(caller)
+        };
+
         let message = Message {
-            caller: call_args.caller,
-            target: call_args.to,
-            gas_limit: call_args.gas,
-            data: call_args.calldata,
+            caller: caller,
```
Comment (bajpai244) [inline crates/evm/src/call_helpers.cairo:90]: we can just do:
`caller,`

File: crates/evm/src/call_helpers.cairo [hunk 4]
```diff
@@ -44,108 +45,57 @@ enum CallType {
...
+            target: to,
+            gas_limit: gas,
+            data: calldata.span(),
             code: code,
```
Comment (bajpai244) [inline crates/evm/src/call_helpers.cairo:94]: same here:

`code,`

File: crates/evm/src/call_helpers.cairo (hunk 1 above)
Comment (enitrat) [reply crates/evm/src/call_helpers.cairo:52]: I don't necessarily agree, if they make sense yes, but in these case ours carry more semantics

File: crates/evm/src/instructions/system_operations.cairo [hunk 5]
```diff
@@ -30,28 +30,70 @@ impl SystemOperations of SystemOperationsTrait {
     /// CALL
     /// # Specification: https://www.evm.codes/#f1?fork=shanghai
     fn exec_call(ref self: VM) -> Result<(), EVMError> {
-        let call_args = self.prepare_call(@CallType::Call)?;
+        let gas = self.stack.pop_u128()?;
+        let to = self.stack.pop_eth_address()?;
+        let value = self.stack.pop()?;
+        let args_offset = self.stack.pop_usize()?;
```
Comment (bajpai244) [inline crates/evm/src/instructions/system_operations.cairo:36]: can be a different issue, but probably we should look to have variable names as close to ELS.

image

File: crates/evm/src/instructions/system_operations.cairo [hunk 6]
```diff
@@ -62,24 +104,92 @@ impl SystemOperations of SystemOperationsTrait {
...
+        let args_size = self.stack.pop_usize()?;
+        let ret_offset = self.stack.pop_usize()?;
+        let ret_size = self.stack.pop_usize()?;
+
+        let args_max_offset = args_offset + args_size;
+        let ret_max_offset = ret_offset + ret_size;
+
+        let to = self.message().target.evm;
+
+        let max_memory_size = if args_max_offset > ret_max_offset {
```
Comment (bajpai244) [inline crates/evm/src/instructions/system_operations.cairo:140]: we can create a seperate issue for this, to model our memory extension gas function to be similar to ELS.

i.e, the function itself can abstract away the logic of figuring out the max memory expansion.
image

File: crates/evm/src/tests/test_call_helpers.cairo
Comment (bajpai244) [inline crates/evm/src/tests/test_call_helpers.cairo:1]: I think we will need replacements for these tests, and I guess we would need them in this PR itself to make sure that we are passing the expected behaviour.

File: crates/evm/src/call_helpers.cairo (hunk 1 above)
Comment (bajpai244) [reply crates/evm/src/call_helpers.cairo:52]: Hmm, I guess something we can discuss later on, but I feel the closer we are to the ELS, the better.

**Review by enitrat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/648)* status: 0 of 6 files reviewed, 7 unresolved discussions (waiting on @bajpai244 and @ClementWalter)

___
*[`crates/evm/src/call_helpers.cairo` line 52 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/648#gh-1424039289:-NlVJcs92uHXp2RMVu7Q:b-r063gf) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f59341c38875053f5c8d91261ba1239004b53400/crates/evm/src/call_helpers.cairo#L52)):*
<details><summary><i>Previously, bajpai244 (Harsh Bajpai) wrote…</i></summary><blockquote>

Hmm, I guess something we can discuss later on, but I feel the closer we are to the ELS, the better.
</blockquote></details>

EELS is poorly designed in a lot of places!
___
*[`crates/evm/src/call_helpers.cairo` line 75 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/648#gh-1424041777:-NlVJfYuBjQ8Peu1dci7:ba3dfob) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f59341c38875053f5c8d91261ba1239004b53400/crates/evm/src/call_helpers.cairo#L75)):*
<details><summary><i>Previously, bajpai244 (Harsh Bajpai) wrote…</i></summary><blockquote>

Is this comment correctly placed?
</blockquote></details>

Removed
___
*[`crates/evm/src/instructions/system_operations.cairo` line 36 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/648#gh-1424098121:-NlVJlAs09wR-YF4dArM:b-j5e57o) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/ad2645676df4a1eb66ac7e19cac73fedccf853d9/crates/evm/src/instructions/system_operations.cairo#L36)):*
<details><summary><i>Previously, bajpai244 (Harsh Bajpai) wrote…</i></summary><blockquote>

can be a different issue, but probably we should look to have variable names as close to ELS.

image

</blockquote></details>

same answer as above
___
*[`crates/evm/src/instructions/system_operations.cairo` line 140 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/648#gh-1424110007:-NlVJo4d1eDJm_4qlCin:b7h4g8y) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/ad2645676df4a1eb66ac7e19cac73fedccf853d9/crates/evm/src/instructions/system_operations.cairo#L140)):*
<details><summary><i>Previously, bajpai244 (Harsh Bajpai) wrote…</i></summary><blockquote>

we can create a seperate issue for this, to model our memory extension gas function to be similar to ELS.

i.e, the function itself can abstract away the logic of figuring out the max memory expansion.
image

</blockquote></details>

Not adapted to Cairo - creating a List of Tuple is overkill
___
*[`crates/evm/src/tests/test_call_helpers.cairo` line 0 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/648#gh-1424116431:-NlVJu4XAlaxp5zZZxVR:b6n50t8) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/ad2645676df4a1eb66ac7e19cac73fedccf853d9/crates/evm/src/tests/test_call_helpers.cairo#L0)):*
<details><summary><i>Previously, bajpai244 (Harsh Bajpai) wrote…</i></summary><blockquote>

I think we will need replacements for these tests, and I guess we would need them in this PR itself to make sure that we are passing the expected behaviour.
</blockquote></details>

Discussed internally - we will rely on EF-Tests mostly. If we want to increase coverage we could eventually add internal functions to test the preparation of the generic call

**Review by bajpai244 — APPROVED** [review]
LGTM 🚀

**Review by bajpai244 — CHANGES_REQUESTED** [review]
@enitrat the tests are failing, we have to make them pass before we can merge.

**Review by bajpai244 — APPROVED** [review]

**Pull request merged**